neutrons/SNAPRed
# A file-by-file rebase onto the 'SNAPRed / next' branch from the 'ewm3702_saving_calibration_records' branch.

**I've kept most of the changes from the original branch, except for the following:**

  * Workspace I/O methods are updated to be consistent with the current service layer.

  * `CalibrationRecord` <workspace list> functionality has been modified to no longer use the proposed `WorkspaceInfo` class.  The information contained in this class was _derivable_ from the workspaces themselves, and so using this class didn't really make much sense.

  * `CalibrationRecord` now uses a **workspace dict** keyed by `WorkspaceNameGenerator` workspace types.  At each key there can be a list containing several workspaces.  This is only "loosely" connected to the WNG-types at the moment, but will serve as a placeholder until this can be done more completely.

  * Arguably, since the workspaces in the `CalibrationRecord` are always the same, the only purpose of storing this information in the record would be if for some reason the end users need to see this information.  We need to check with Malcolm about this point.  If this information is unnecessary, the actual list should be moved to `application.yml`, and be by WNG-type.

  * All unit tests are presently passing.

  * All UI and Filesystem functionality of the original story is functioning as expected.

## Description of work

A calibration record file has a list of workspaces that are saved as files in the same directory.  The names of the workspaces and their filenames must meet certain criteria:

- The name should end with a run number followed by a version number, for example
`diffract_consts_046680_v0015` for a workspace and
`diffract_consts_046680_v0015.h5` for the corresponding file
 - Run numbers and version numbers should be zero-padded as shown in the example above
 - Workspace names should be easy to interpret. For example, instead of the current `difc`, use `diffract_consts`

The above rules also apply to the calibration **metric** workspaces.  The metric workspaces are not saved as files, but are listed in the Workspace window of the SNAPRed UI.

The names of calibration version directories should also be zero-padded, e.g. `v_0015` instead of the current `v_15`.

Filename extensions must reflect the corresponding workspace Mantid types.

Calibration and Normalization records must be saved in separate directories:
```
../Calibration/Powder/04bd2c53f6bf6754/diffraction
../Calibration/Powder/04bd2c53f6bf6754/normalization

```
**ATTENTION:**

This PR also fixes a UI bug introduced by [PR#288](https://github.com/neutrons/SNAPRed/pull/228): on the Assessing tab the `CalibrationIndex` drop-down list  is always empty. To fix the bug, a missing call to `CalibrationAssesmentView` `updateRunNumber` was re-instated:
``` python
   def _triggerCalibrationReduction(self, workflowPresenter):
    ....
        self._calibrationAssessmentView.updateRunNumber(self.runNumber)
```
Correspondingly, the implementation of `updateRunNumber` was also re-instated.

**KNOWN ISSUES**

- If the workspace name contains an iteration number, the number is positioned after the run number. It should be positioned before the run number.
- Before saving a workspace the iteration number needs to be removed from the workspace name.
- Because this does _not_ occur, although multiple iterations will execute without any issues,  they cannot be correctly reloaded by the "assessment" panel.
- Correct implementation of this feature probably requires a WNG-name parser.
- **THIS MAY STILL NEED SOME WORK**, in another story...

## Explanation of work

For the purposes of `WorkspaceNameGenerator` (WNG), all relevant name templates have been updated to make sure that the run number and version tokens are positioned as required. Also, token name changes like `difc` -> `diffract_consts` were made.

Zero-padding of run number and version number is accomplished with the help of a new class `ValueFormatter`.
Correspondingly, a one-line change to `NameBuilder` `__getattr__` method:

  ``` python
    def __getattr__(self, key):
        if key not in self.keys:
            raise RuntimeError(f"Key [{key}] not a valid property for given name.")

        def setValue(value):
            
             # FORMAT THE VALUE
             value = ValueFormatter.formatValueByKey(key, value)
            
            self.props[key] = value
            return self
  ```
allows for **automatic** formatting of the corresponding name tokens without any changes to the client code. For example, an existent line of code 
``` python
wng.diffCalInput().runNumber(runId).build()
```
will result in automatic formatting of `runId`.

Filenames, unlike workspace names, might not be built exclusively with WNG: a version suffix (as well as a filename extension) may be added "by hand" immediately before saving the workspace. To format the version number in such cases a call is placed directly to `ValueFormater`, e.g.
``` python
        if version:
            filePath += "_" + ValueFormatter.formatVersion(version)
            filePath += ".h5"
 
```
The same applies to formatting a calibration version directory name, e.g.
``` python
   def _constructCalibrationDataPath(self, runId: str, version: str):
        stateId, _ = self._generateStateId(runId)
        statePath = self._constructCalibrationStatePath(stateId)
        calibrationVersionPath: str = statePath + "v_{}/".format(
            ValueFormatter.formatVersion(version=version, use_v_prefix=False)
        )
        return calibrationVersionPath
```

A new name template was added to WNG to account for metric workspaces that are not assigned a version yet, but have a timestamp instead: `workspace:nameTemplate:timed_metric`.

## To test

Set up a directory structure:
../Calibration/Powder

Set up a directory structure:
../Calibration/Powder

For testing diffraction calibration use the following parameters:
Run Number: 46680
Convergence Threshold: 0.1
Peak Intensity Threshold: 0.0001
Bins Across Peak Width: 10
Sample: Diamond_001.json
Grouping File: Column lite

For testing normalization calibration use the following parameters:
Run Number: 58882
Background Number: 58813
Smoothing Parameter: 0.000001
Sample: Silicon_NIST_640D_001
Grouping File: Column lite

TEST 1

1. From the `Test Panel` `Calibrating` tab run Diffraction Calibration
2. When `Assessing` tab is shown, click Continue
3. On the `Saving` tab fill in the fields and click Continue
4. Check that the saved `CalibrationRecord.json` has the correct list of persistent workspaces.  Check that the corresponding filenames are also correct.

Close and reopen UI

TEST 2
1. From the `Test Panel` `Diffraction Calibration`/`Calibrating` tab run diffraction calibration
2. When `Assessing` tab is shown, verify that version 2 calibration is present in the drop-down list
3. Click Load. Check the Workspaces window to make sure that all expected workspaces have been created

Close and reopen UI

TEST 3
1. From the `Test Panel` `Normalization Calibration`/`Normalization Calibration` tab run normalization calibration.
2. When `Tweak Parameters` tab is shown, click Continue
3.  On the `Saving` tab fill in the fields and click Continue
4. Check that the saved `NormalizationRecord.json` has the correct list of persistent workspaces.  Check that the corresponding filenames are also correct.



## Link to EWM item


[EWM#3702](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=3702)
